KIC 2.6: Clarification for upgrading to Kong Gateway 3.x (#4470)
Kong/docs.konghq.com: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md

Ingress rule paths have no way to indicate that a path is a regular expression.
The `ImplementationSpecific` path type can contain either regular expression or
non-regular expression paths. If you have existing Ingresses with regular

> expression paths, those paths will break if you upgrade
> to 3.x without updating configuration.

Reviewer: Slotting the additional warning from #4471 in here arbitrarily. It's not directly related to any upgrade concept, but I want to remind about this because we see this misconfiguration occasionally, and it'll cause worse problems with 3.x.

```suggestion
expression paths, those paths will break if you upgrade
to 3.x without updating configuration.

{:.important}
> **Important**: `Prefix` and `Exact` rules [must never use regular expressions](/kubernetes-ingress-controller/{{page.kong_version}}/concepts/ingress-versions/#networkingk8siov1).
Only use regular expressions in `ImplementationSpecific` rules.
```

Ingress rule paths have no way to indicate that a path is a regular expression.
The `ImplementationSpecific` path type can contain either regular expression or
non-regular expression paths. If you have existing Ingresses with regular
expression paths, those paths will break if you upgrade
to 3.x without updating configuration.

{:.important}
> **Important**: `Prefix` and `Exact` rules [must never use regular expressions](/kubernetes-ingress-controller/{{page.kong_version}}/concepts/ingress-versions/#networkingk8siov1).
Only use regular expressions in `ImplementationSpecific` rules.

The new 3.x paths are also incompatible with 2.x. Adding the `~` character to

> Ingress rules directly would break incremental upgrades. To smooth the migration

Reviewer 2: ```suggestion
Ingress rules directly breaks incremental upgrades. To smooth the migration
```

replace the `kong` Ingress class name and `kong` namespace with values
appropriate to your environment.

After you have upgraded to Kong 3.x, update all of your regular expression paths
to include the `~` prefix in the Ingress itself at your earliest convenience.
The heuristic is only intended for decreasing downtime during upgrades, so
it behaves exactly like the version included in Kong 2.8.x, bugs
included. Once you added the prefix to all regular expression paths, you can
disable `enableLegacyRegexDetection`.

{:.important}
> **Important**: The `enableLegacyRegexDetection` option is meant to be **temporary**.

> Use it for migration only.

Reviewer 2: ```suggestion
Only use it for migration.
```
